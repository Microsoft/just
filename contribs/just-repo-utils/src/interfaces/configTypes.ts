/**
 * defers the actual parsing and loading of a JSON file behind a loader function.  Generally the
 * existence check will already be completed such that a successful return can be guaranteed
 */
export type ConfigLoader<T> = () => T;

/**
 * Rush project info, corresponds to each project within the larger rush repo
 */
export interface RushProject {
  packageName: string;
  projectFolder: string;
  shouldPublish?: boolean;
  versionPolicyName?: string;
}

/**
 * Root of the parsed rush json
 */
export interface RushJson {
  projects: RushProject[];
  rushVersion: string;
  // more properties can be added as needed
}

/**
 * Loader type for deferred loading of rush JSON
 */
export type RushJsonLoader = ConfigLoader<RushJson>;

/**
 * Format for lerna json, packages reflect the globs denoting sub-packages
 */
export interface LernaJson {
  packages: string[];
  useWorkspaces?: boolean;
  npmClient?: string;
  version?: string;
}

/**
 * Loader type for deferred loading of lerna json
 */
export type LernaJsonLoader = ConfigLoader<LernaJson>;

interface Dependencies {
  [key: string]: string;
}

export interface PackageJson {
  name: string;
  description?: string;
  dependencies?: Dependencies;
  devDependencies?: Dependencies;
  peerDependencies?: Dependencies;
  keywords?: string;
  just?: {
    /** Indicator that this package should be treated as the root */
    root?: boolean;

    /** Stack that the package is tracking */
    stack?: string;
  };
  [key: string]: any;
}

/**
 * loader type for deferred loading of package JSON
 */
export type PackageJsonLoader = ConfigLoader<PackageJson>;
