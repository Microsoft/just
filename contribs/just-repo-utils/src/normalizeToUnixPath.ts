import path from 'path';

/**
 * take a path, call path.normalize, then make sure it uses forward slashes
 * @param base - path to put into forward slashed form
 */
export function normalizeToUnixPath(base: string): string {
  return path.normalize(base).replace(/\\/g, '/');
}
