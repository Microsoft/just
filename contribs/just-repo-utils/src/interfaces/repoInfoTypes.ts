import { PackageJsonLoader, RushJsonLoader, LernaJsonLoader } from './configTypes';

/**
 * callback function for walking to the root.  Returning true will cancel the walk at
 * that point, a false or no return result will keep walking.
 */
export type FindRootCallback = (cwd: string) => boolean | void;

/**
 * types of supported monorepos
 */
export type MonorepoType = 'lerna' | 'rush';

/**
 * Info for the repository
 */
export interface RepoInfo {
  /** root of the enlistment */
  rootPath: string;

  /** loader for package JSON */
  getPackageJson: PackageJsonLoader;

  /** if it is a monorepo denotes whether it is rush or lerna based */
  monorepo?: MonorepoType;

  /** loader function to get the parsed rush json */
  getRushJson?: RushJsonLoader;

  /** loader function to get the parsed lerna json */
  getLernaJson?: LernaJsonLoader;
}

/**
 * Options for querying repository info
 */
export interface RepoInfoOptions {
  /** current working directory to start from */
  cwd?: string;
}
