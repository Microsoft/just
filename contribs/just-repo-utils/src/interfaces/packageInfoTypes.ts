import { PackageJsonLoader } from './configTypes';
import { RepoInfoOptions } from './repoInfoTypes';

export interface PackageEntry {
  path: string;
  getConfig: PackageJsonLoader;
  dependencies: { [key: string]: PackageEntry };
}

export interface PackageEntries {
  [key: string]: PackageEntry;
}

/**
 * cache strategy to use for package info.  Values are:
 *  normal - try to load from the cache and save normally
 *  update - don't load from the cache and save a new copy to the cache
 *  no-cache - don't load or save to the cache
 */
export type CacheStrategy = 'normal' | 'no-cache' | 'update';

/**
 * optional options for package info routines
 */
export interface PackageInfoOptions extends RepoInfoOptions {
  /**
   * caching strategy to use for package info queries, defaults to 'normal'
   */
  strategy?: CacheStrategy;

  /**
   * when querying dependencies this is the target package name to use for the query.  If
   * omitted this is assumed to be the package in process.cwd()
   */
  target?: string;

  /**
   * normally a package is a dependency if it is in dependencies or devDependencies as these
   * two are effectively the same from a repository package install perspective.  (i.e. changing
   * from one to the other won't change the yarn lockfile)
   *
   * Specifying one of these will narrow the returned results
   */
  dependencyType?: 'dependencies' | 'devDependencies';
}

export interface PackageInfo {
  /**
   * Retrieve an array of paths for the entries
   */
  paths: () => string[];

  /**
   * Retrieve an array of names for the entries
   */
  names: () => string[];

  /**
   * Return a filtered PackageInfo for either targeting the name of a package or the
   * package in the current working directory
   */
  dependencies: (options?: PackageInfoOptions) => PackageInfo;

  /**
   * Raw entry table for clients who want to do name to path mappings or have access to the package
   * json files for the packages
   */
  entries: PackageEntries;
}
