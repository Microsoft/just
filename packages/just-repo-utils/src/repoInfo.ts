import { getConfigLoader, loadCJson } from './readConfigs';
import { LernaJson, RushJson, PackageJson, PackageJsonLoader } from './interfaces/configTypes';
import { RepoInfo, FindRootCallback, RepoInfoOptions } from './interfaces/repoInfoTypes';
import path from 'path';
import fse from 'fs-extra';

let _repoInfo: RepoInfo | undefined = undefined;

/**
 * A more comprehensive check for repo root.  This will check the just section of package.json to
 * see if this has been flagged as a root or if it is using the monorepo stack.  It will also check
 * for rush.json or lerna.json.  Any of these should be treated as the repository root.
 *
 * @param cwd - current working directory to test
 * @param loader - loader function for
 */
function isRepoRoot(cwd: string, config?: PackageJson): boolean | undefined {
  const just = config && config.just;
  const isRootFromJust = just && (just.root || (just.stack && just.stack === 'just-stack-monorepo'));
  const isRootFromMonorepoConfigs = config && (fse.existsSync(path.join(cwd, 'rush.json')) || fse.existsSync(path.join(cwd, 'lerna.json')));
  return isRootFromJust || isRootFromMonorepoConfigs;
}

/**
 * Finds the root of the git repository, i.e. where .git resides
 *
 * @param cb - callback function to execute at each level.  A true result for the callback will
 * cancel the walk and return the current path at the time it was cancelled.
 */
export function findGitRoot(cb?: FindRootCallback, options?: RepoInfoOptions): string {
  let cwd = (options && options.cwd) || process.cwd();
  const root = path.parse(cwd).root;

  while (cwd !== root) {
    if ((cb && cb(cwd)) || fse.existsSync(path.join(cwd, '.git'))) {
      return cwd;
    }
    cwd = path.dirname(cwd);
  }
  throw 'No repository root found!';
}

/**
 * helper for finding the repository root that also returns the package loader so that it can be
 * retained for use later.
 *
 * @param cb - standard findGitRoot/findRepoRoot callback
 */
function findRepoRootWithConfig(cb?: FindRootCallback, options?: RepoInfoOptions): [string, PackageJsonLoader | undefined] {
  let loader: PackageJsonLoader | undefined = undefined;
  const path = findGitRoot(cwd => {
    loader = getConfigLoader<PackageJson>(cwd, 'package.json');
    return (cb && cb(cwd)) || isRepoRoot(cwd, loader && loader());
  }, options);
  return [path, loader];
}

/**
 * Finds the root of the repository, will handle various end conditions such as git root,
 * rush/lerna config existence, or certain just settings in package.json
 *
 * @param cb - callback function of the same type as for findGitRoot
 */
export function findRepoRoot(cb?: FindRootCallback, options?: RepoInfoOptions): string {
  return findRepoRootWithConfig(cb, options)[0];
}

/**
 * Retrieve info for the repository.  This will walk up from the current working directory
 * until it finds the git root and then prepare loaders for various monorepo config files.
 *
 * Note that this uses in-module caching to avoid traversing unnecessarily.
 */
export function getRepoInfo(options?: RepoInfoOptions): RepoInfo {
  if (_repoInfo && (!options || !options.cwd)) {
    return _repoInfo;
  }

  const [rootPath, packageLoader] = findRepoRootWithConfig(undefined, options);
  const getRushJson = getConfigLoader<RushJson>(rootPath, 'rush.json', loadCJson);
  const getLernaJson = getConfigLoader<LernaJson>(rootPath, 'lerna.json');
  const isMonoRepo = getRushJson || getLernaJson;
  _repoInfo = {
    rootPath: rootPath,
    getRushJson,
    getLernaJson,
    getPackageJson: packageLoader!,
    ...(isMonoRepo && { monorepo: getRushJson ? 'rush' : 'lerna' })
  };
  return _repoInfo;
}
