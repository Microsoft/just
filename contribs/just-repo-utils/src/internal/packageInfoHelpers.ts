import { PackageJson, PackageJsonLoader, RushProject } from '../interfaces/configTypes';
import { PackageEntries, PackageEntry, PackageInfo, PackageInfoOptions } from '../interfaces/packageInfoTypes';
import { getConfigLoader, readPackageJson } from '../readConfigs';
import { normalizeToUnixPath } from '../normalizeToUnixPath';
import path from 'path';
import glob from 'glob';

function buildPackageInfo(root: string, pkgPath: string, pkgJsonName?: string): PackageEntries {
  const pkgJsonPath = pkgJsonName ? path.join(root, pkgPath, pkgJsonName) : path.join(root, pkgPath);
  const getConfig = getConfigLoader<PackageJson>(pkgJsonPath) as PackageJsonLoader;
  return {
    [getConfig().name]: {
      path: normalizeToUnixPath(path.dirname(pkgJsonPath)),
      getConfig,
      dependencies: {}
    }
  };
}

function buildPackageInfoForGlob(root: string, pkgGlob: string): PackageEntries {
  const matchPattern = pkgGlob + '/package.json';
  const globOptions = { cwd: root, ignore: '**/node_modules/**' };
  return Object.assign({}, ...glob.sync(matchPattern, globOptions).map(subPath => buildPackageInfo(root, subPath)));
}

function addPackageDependencyBranch(repoInfo: PackageEntries, pkg: PackageEntry, key: string): void {
  const config = pkg.getConfig();
  const section = config[key];
  const dependencies = pkg.dependencies;
  if (section) {
    Object.keys(section).forEach(dependency => {
      if (repoInfo[dependency] && !dependencies[dependency]) {
        dependencies[dependency] = repoInfo[dependency];
      }
    });
  }
}

function addPackageDependencies(info: PackageEntries): void {
  Object.keys(info).forEach(pkg => {
    addPackageDependencyBranch(info, info[pkg], 'dependencies');
    addPackageDependencyBranch(info, info[pkg], 'devDependencies');
  });
}

export function buildPackageInfoFromGlobs(root: string, globs: string[]): PackageEntries {
  const results: PackageEntries = Object.assign({}, ...globs.map(glob => buildPackageInfoForGlob(root, glob)));
  addPackageDependencies(results);
  return results;
}

export function buildPackageInfoFromRushProjects(root: string, projects: RushProject[]): PackageEntries {
  const results: PackageEntries = Object.assign(
    {},
    ...projects.map(project => buildPackageInfo(root, project.projectFolder, 'package.json'))
  );
  addPackageDependencies(results);
  return results;
}

function findRecursiveDependencies(collector: PackageEntries, entry: PackageEntry, depType?: PackageInfoOptions['dependencyType']): void {
  const dependencies = entry.dependencies;
  Object.keys(dependencies).forEach(dep => {
    if (!collector[dep]) {
      const configDeps = depType && entry.getConfig()[depType];
      if (!depType || (configDeps && configDeps[dep])) {
        collector[dep] = dependencies[dep];
        findRecursiveDependencies(collector, dependencies[dep], depType);
      }
    }
  });
}

export function infoFromEntries(entries: PackageEntries): PackageInfo {
  return {
    paths: () => Object.keys(entries).map(pkgName => normalizeToUnixPath(entries[pkgName].path)),
    names: () => Object.keys(entries),
    dependencies: (options?: PackageInfoOptions) => {
      const target = (options && options.target) || readPackageJson(process.cwd())!.name;
      const baseEntry = entries[target];
      const collector: PackageEntries = {};
      if (baseEntry) {
        findRecursiveDependencies(collector, baseEntry, options ? options.dependencyType : undefined);
      }
      return infoFromEntries(collector);
    },
    entries
  };
}
