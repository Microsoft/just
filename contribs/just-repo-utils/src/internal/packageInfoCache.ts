import { getRepoInfo } from '../repoInfo';
import { getRepoHashKey, getCachePath } from '../cacheUtils';
import { PackageEntries } from '../interfaces/packageInfoTypes';
import { getConfigLoader, readJsonConfig } from '../readConfigs';
import { PackageJson, PackageJsonLoader } from '../interfaces/configTypes';
import fs from 'fs-extra';
import path from 'path';

interface ISerializablePackageEntry {
  path: string;
  dependencies: string[];
}

interface ISerializableRepoPackages {
  [key: string]: ISerializablePackageEntry;
}

interface IPackageInfoCacheJson {
  hash: string;
  packages: ISerializableRepoPackages;
}

const cacheFileName = 'package-info.json';
const cachePath = getCachePath(cacheFileName, true);

/**
 * @private only exposed for testing
 */
export function getSerializableRepoPackages(info: PackageEntries): ISerializableRepoPackages {
  const results: ISerializableRepoPackages = {};
  Object.keys(info).forEach(pkgName => {
    results[pkgName] = {
      path: info[pkgName].path,
      dependencies: Object.keys(info[pkgName].dependencies)
    };
  });
  return results;
}

/**
 * @private only exposed for testing
 */
export function getRepoPackagesFromSerializableForm(info: ISerializableRepoPackages): PackageEntries {
  const results: PackageEntries = {};
  // build the initial set
  Object.keys(info).forEach(pkgName => {
    const entry = info[pkgName];
    results[pkgName] = {
      path: entry.path,
      getConfig: getConfigLoader<PackageJson>(entry.path, 'package.json') as PackageJsonLoader,
      dependencies: {}
    };
  });
  // now link dependencies
  Object.keys(info).forEach(pkgName => {
    const dependencies = results[pkgName].dependencies;
    info[pkgName].dependencies.forEach(pkg => {
      dependencies[pkg] = results[pkg];
    });
  });
  return results;
}

/**
 * this saves the loaded package info into a JSON cache file
 * @param pkgInfo - package info to save in a cache file
 */
export function cachePackageInfo(pkgInfo: PackageEntries): void {
  const repo = getRepoInfo();
  const toJson: IPackageInfoCacheJson = {
    hash: getRepoHashKey(repo.rootPath),
    packages: getSerializableRepoPackages(pkgInfo)
  };
  // ensure the directory is created
  const cacheDir = path.dirname(cachePath);
  if (!fs.pathExistsSync(cacheDir)) {
    fs.mkdirpSync(cacheDir);
  }
  // now write out the cache file
  fs.writeFileSync(cachePath, JSON.stringify(toJson, null, 2));
}

/**
 * this attempts to load package info from a JSON cache file
 */
export function retrievePackageInfo(): PackageEntries | undefined {
  const repo = getRepoInfo();
  const jsonData = readJsonConfig<IPackageInfoCacheJson>(cachePath);
  if (jsonData && jsonData.hash == getRepoHashKey(repo.rootPath)) {
    return getRepoPackagesFromSerializableForm(jsonData.packages);
  }
  return undefined;
}
