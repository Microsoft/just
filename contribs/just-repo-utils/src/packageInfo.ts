import { PackageInfoOptions, PackageInfo } from './interfaces/packageInfoTypes';
import { getRepoInfo } from './repoInfo';

function validateOptions(options?: PackageInfoOptions): PackageInfoOptions {
  options = options || {};
  // if a current working directory is being specified the primary use case is querying dependencies or packages of a different
  // repository.  In this case, any information retrieved shouldn't be stored in our cache as it may or may not be relevant
  if (options.cwd) {
    options.strategy = 'no-cache';
  }
  return options;
}

/**
 * retrieves information about the packages in the repository
 * @param strategy - cache strategy to use for loading, defaults to normal
 */
export function getPackageInfo(options?: PackageInfoOptions): PackageInfo {
  options = validateOptions(options);
  const { retrievePackageInfo, cachePackageInfo } = require('./internal/packageInfoCache');
  const { infoFromEntries, buildPackageInfoFromGlobs, buildPackageInfoFromRushProjects } = require('./internal/packageInfoHelpers');
  const { strategy = 'normal' } = options;
  let repoPackageInfo = strategy === 'normal' && retrievePackageInfo();
  if (!repoPackageInfo) {
    const repo = getRepoInfo(options);
    if (repo.getLernaJson) {
      repoPackageInfo = buildPackageInfoFromGlobs(repo.rootPath, repo.getLernaJson().packages);
    } else if (repo.getRushJson) {
      repoPackageInfo = buildPackageInfoFromRushProjects(repo.rootPath, repo.getRushJson().projects);
    }
    if (strategy !== 'no-cache' && repoPackageInfo) {
      cachePackageInfo(repoPackageInfo);
    }
  }
  return infoFromEntries(repoPackageInfo || {});
}

export function getDependentPackageInfo(options?: PackageInfoOptions): PackageInfo {
  options = options || {};
  return getPackageInfo(options).dependencies(options);
}
