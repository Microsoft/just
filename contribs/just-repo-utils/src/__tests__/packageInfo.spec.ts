import { PackageEntries } from '../interfaces/packageInfoTypes';
import { getPackageInfo } from '../packageInfo';
import { getRepoPackagesFromSerializableForm, getSerializableRepoPackages } from '../internal/packageInfoCache';
import { findGitRoot } from '../repoInfo';
import { normalizeToUnixPath } from '../normalizeToUnixPath';

const _rootPath = normalizeToUnixPath(findGitRoot());

function trimPath(rootPath: string, fullPath: string): string {
  const index = fullPath.indexOf(rootPath);
  return index >= 0 ? fullPath.substr(index + rootPath.length) : fullPath;
}

/**
 * create a version of PackageEntries that does not have config loaders so that they can be compared cleanly
 * @param entries - entries to create a comparable copy of
 */
function comparableEntries(entries: PackageEntries): PackageEntries {
  return Object.assign(
    {},
    ...Object.keys(entries)
      .sort()
      .map(name => {
        const entry = entries[name];
        return {
          [name]: {
            path: trimPath(_rootPath, entry.path),
            dependencies: Object.assign(
              {},
              Object.keys(entry.dependencies).map(name => ({ [name]: {} }))
            )
          }
        };
      })
  );
}

describe('packageInfo', () => {
  test('load with no-cache', () => {
    const info = getPackageInfo({ strategy: 'no-cache' });
    Object.keys(info.entries).forEach(name => {
      const entry = info.entries[name];
      const config = entry.getConfig();
      expect(config.name).toEqual(name);
    });
  });

  test('load with cache', () => {
    const info = getPackageInfo({ strategy: 'normal' });
    Object.keys(info.entries).forEach(name => {
      const entry = info.entries[name];
      const config = entry.getConfig();
      expect(config.name).toEqual(name);
    });
  });

  test('serialize and back are equivalent', () => {
    const info = getPackageInfo();
    const serializedForm = getSerializableRepoPackages(info.entries);
    const unserializedEntries = getRepoPackagesFromSerializableForm(serializedForm);
    const reserializedForm = getSerializableRepoPackages(unserializedEntries);
    expect(serializedForm).toEqual(reserializedForm);
    expect(comparableEntries(info.entries)).toEqual(comparableEntries(unserializedEntries));
  });

  test('cached and non-cached are equivalent', () => {
    const infoCached = getPackageInfo();
    const infoDirect = getPackageInfo({ strategy: 'no-cache' });
    expect(comparableEntries(infoCached.entries)).toEqual(comparableEntries(infoDirect.entries));
  });

  test('paths', () => {
    const paths = getPackageInfo()
      .paths()
      .map(path => trimPath(_rootPath, path));
    expect(paths).toMatchSnapshot();
  });

  test('names', () => {
    const names = getPackageInfo().names();
    expect(names).toMatchSnapshot();
  });

  test('info matches snapshot', () => {
    const info = getPackageInfo();
    expect(comparableEntries(info.entries)).toMatchSnapshot();
  });

  test('dependency graph', () => {
    const info = getPackageInfo().dependencies({ target: 'just-scripts' });
    expect(comparableEntries(info.entries)).toMatchSnapshot();
  });

  test('dependency graph with filters', () => {
    const info = getPackageInfo().dependencies({ target: 'just-scripts', dependencyType: 'dependencies' });
    expect(comparableEntries(info.entries)).toMatchSnapshot();
  });

  test('dependency graph with devDependency filter', () => {
    const info = getPackageInfo().dependencies({ target: 'just-scripts', dependencyType: 'devDependencies' });
    expect(comparableEntries(info.entries)).toMatchSnapshot();
  });
});
