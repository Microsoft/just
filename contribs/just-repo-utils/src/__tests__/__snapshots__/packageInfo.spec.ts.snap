// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`packageInfo dependency graph 1`] = `
Object {
  "just-scripts-utils": Object {
    "dependencies": Object {
      "0": Object {
        "just-task-logger": Object {},
      },
    },
    "path": "/packages/just-scripts-utils",
  },
  "just-task": Object {
    "dependencies": Object {
      "0": Object {
        "just-task-logger": Object {},
      },
    },
    "path": "/packages/just-task",
  },
  "just-task-logger": Object {
    "dependencies": Object {},
    "path": "/packages/just-task-logger",
  },
}
`;

exports[`packageInfo dependency graph with devDependency filter 1`] = `Object {}`;

exports[`packageInfo dependency graph with filters 1`] = `
Object {
  "just-scripts-utils": Object {
    "dependencies": Object {
      "0": Object {
        "just-task-logger": Object {},
      },
    },
    "path": "/packages/just-scripts-utils",
  },
  "just-task": Object {
    "dependencies": Object {
      "0": Object {
        "just-task-logger": Object {},
      },
    },
    "path": "/packages/just-task",
  },
  "just-task-logger": Object {
    "dependencies": Object {},
    "path": "/packages/just-task-logger",
  },
}
`;

exports[`packageInfo info matches snapshot 1`] = `
Object {
  "create-just": Object {
    "dependencies": Object {
      "0": Object {
        "just-plop-helpers": Object {},
      },
      "1": Object {
        "just-scripts-utils": Object {},
      },
    },
    "path": "/packages/create-just",
  },
  "create-uifabric": Object {
    "dependencies": Object {
      "0": Object {
        "create-just": Object {},
      },
    },
    "path": "/packages/create-uifabric",
  },
  "example-lib": Object {
    "dependencies": Object {
      "0": Object {
        "just-scripts": Object {},
      },
    },
    "path": "/packages/example-lib",
  },
  "just-plop-helpers": Object {
    "dependencies": Object {},
    "path": "/packages/just-plop-helpers",
  },
  "just-repo-utils": Object {
    "dependencies": Object {
      "0": Object {
        "just-task-logger": Object {},
      },
    },
    "path": "/packages/just-repo-utils",
  },
  "just-scripts": Object {
    "dependencies": Object {
      "0": Object {
        "just-scripts-utils": Object {},
      },
      "1": Object {
        "just-task": Object {},
      },
    },
    "path": "/packages/just-scripts",
  },
  "just-scripts-utils": Object {
    "dependencies": Object {
      "0": Object {
        "just-task-logger": Object {},
      },
    },
    "path": "/packages/just-scripts-utils",
  },
  "just-stack-monorepo": Object {
    "dependencies": Object {
      "0": Object {
        "just-plop-helpers": Object {},
      },
    },
    "path": "/packages/just-stack-monorepo",
  },
  "just-stack-react": Object {
    "dependencies": Object {
      "0": Object {
        "just-plop-helpers": Object {},
      },
    },
    "path": "/packages/just-stack-react",
  },
  "just-stack-single-lib": Object {
    "dependencies": Object {
      "0": Object {
        "just-plop-helpers": Object {},
      },
      "1": Object {
        "just-scripts": Object {},
      },
    },
    "path": "/packages/just-stack-single-lib",
  },
  "just-stack-uifabric": Object {
    "dependencies": Object {},
    "path": "/packages/just-stack-uifabric",
  },
  "just-task": Object {
    "dependencies": Object {
      "0": Object {
        "just-task-logger": Object {},
      },
    },
    "path": "/packages/just-task",
  },
  "just-task-docs": Object {
    "dependencies": Object {},
    "path": "/packages/documentation/website",
  },
  "just-task-logger": Object {
    "dependencies": Object {},
    "path": "/packages/just-task-logger",
  },
  "just-task-scripts": Object {
    "dependencies": Object {},
    "path": "/scripts",
  },
}
`;

exports[`packageInfo names 1`] = `
Array [
  "create-just",
  "create-uifabric",
  "example-lib",
  "just-plop-helpers",
  "just-repo-utils",
  "just-scripts-utils",
  "just-scripts",
  "just-stack-monorepo",
  "just-stack-react",
  "just-stack-single-lib",
  "just-stack-uifabric",
  "just-task-logger",
  "just-task",
  "just-task-docs",
  "just-task-scripts",
]
`;

exports[`packageInfo paths 1`] = `
Array [
  "/packages/create-just",
  "/packages/create-uifabric",
  "/packages/example-lib",
  "/packages/just-plop-helpers",
  "/packages/just-repo-utils",
  "/packages/just-scripts-utils",
  "/packages/just-scripts",
  "/packages/just-stack-monorepo",
  "/packages/just-stack-react",
  "/packages/just-stack-single-lib",
  "/packages/just-stack-uifabric",
  "/packages/just-task-logger",
  "/packages/just-task",
  "/packages/documentation/website",
  "/scripts",
]
`;
